### PR TITLE
Add support for multiple schedules

### PR DESCRIPTION
**Pull request by Teots**

### CODE REVIEW

**Comment by kensipe** [comment]
@Teots just a quick test of the API seems to indicate something missing... I didn't dig deep yet.  Here the logs:

```[debug] d.m.b.i.BehaviorImpl - Create new count for method: updateJobSpec exception: java.lang.IllegalArgumentException in class dcos.metronome.jobspec.impl.JobSpecServiceDelegate
[error] d.m.j.i.JobSpecPersistenceActor - Repository change failed
java.lang.IllegalArgumentException: requirement failed: Only one schedule supported at the moment
	at scala.Predef$.require(Predef.scala:224)
	at dcos.metronome.api.v1.controllers.JobScheduleController$$anonfun$createSchedule$1.dcos$metronome$api$v1$controllers$JobScheduleController$$anonfun$$addSchedule$1(JobScheduleController.scala:44)
	at dcos.metronome.api.v1.controllers.JobScheduleController$$anonfun$createSchedule$1$$anonfun$apply$8$$anonfun$apply$9.apply(JobScheduleController.scala:48)
	at dcos.metronome.api.v1.controllers.JobScheduleController$$anonfun$createSchedule$1$$anonfun$apply$8$$anonfun$apply$9.apply(JobScheduleController.scala:48)
	at dcos.metronome.repository.impl.kv.KeyValueRepository.dcos$metronome$repository$impl$kv$KeyValueRepository$$updateEntity$1(KeyValueRepository.scala:26)
	at dcos.metronome.repository.impl.kv.KeyValueRepository$$anonfun$update$1.apply(KeyValueRepository.scala:33)
	at dcos.metronome.repository.impl.kv.KeyValueRepository$$anonfun$update$1.apply(KeyValueRepository.scala:32)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:253)
	at scala.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:251)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
```

**Comment by kensipe** [comment]
also... This was 2 schedules added separately on a job and a failure on the second post against the endpoint.

**Comment by kensipe** [comment]
@Teots 
if you remove https://github.com/dcos/metronome/blob/master/api/src/main/scala/dcos/metronome/api/v1/controllers/JobScheduleController.scala#L44

a quick look shows that it works, however then there is 1 test that fails.   Remove that line, update the tests (probably add another test for multiple schedules at the controller level) and I will test and merge.

**Comment by Teots** [comment]
@kensipe I removed this line and update the now failing test. Instead of checking that only one schedule per job is allowed, it now checks that multiple schedules per job are allowed.

**Comment by kensipe** [comment]
@Teots could you rebase please?   I'll take a look shortly when you do.

**Comment by Teots** [comment]
@kensipe I rebased my PR.

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 1]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
   import context.dispatcher
 
   private[impl] var spec = initSpec
-  private[impl] var nextSchedule: Option[Cancellable] = None
-  private[impl] var scheduledAt: Option[DateTime] = None
+  private[impl] var schedules = TrieMap.empty[ScheduleSpec, ScheduleActions]
 
   override def preStart(): Unit = {
-    scheduleNextRun()
+    spec.schedules.foreach(scheduleNextRun(_))
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: not sure why this isn't just `spec.schedules.foreach(scheduleNextRun)`

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 2]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
   }
 
   override def postStop(): Unit = {
-    cancelSchedule()
+    spec.schedules.foreach(cancelSchedule(_))
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: same as above for `cancelSchedule`

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 3]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
   }
 
   override def receive: Receive = around {
-    case StartJob               => runJob()
+    case StartJob(s)            => runJob(s)
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: seems like we could be more descriptive than `s`

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 4]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
-    scheduledAt = None
-    scheduleNextRun()
+
+    // Identify deleted, new and changed schedules
+    val deletedSchedules = schedules.keys.filterNot(newSpec.schedules.contains(_)).toSet
+    val newSchedules = newSpec.schedules.filterNot(schedules.keys.toSet)
+    val updateableSchedules = schedules.keys.filter(newSpec.schedules.contains(_)).toSet
+
+    // Cancel all removed schedules
+    deletedSchedules.foreach(cancelSchedule(_))
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: same as above:  `deletedSchedules.foreach(cancelSchedule)`
and ditto below

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 5]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
+    log.info(s"Start next run of schedule ${s.id} of job ${spec.id}, which was scheduled for ${schedules.get(s).map(_.scheduledAt)}")
     runService.startJobRun(spec)
-    scheduleNextRun()
+    scheduleNextRun(s)
   }
 
-  def scheduleNextRun(): Unit = {
+  def scheduleNextRun(s: ScheduleSpec): Unit = {
+    val scheduleActionsOpt = schedules.get(s)
+    val c = scheduleActionsOpt.map(_.cancellable)
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: not liking `c` as a variable... in this case is isn't used.  lets remove.

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 6]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
-    log.info(s"Start next run of job ${spec.id}, which was scheduled for $scheduledAt")
+  def runJob(s: ScheduleSpec): Unit = {
+    log.info(s"Start next run of schedule ${s.id} of job ${spec.id}, which was scheduled for ${schedules.get(s).map(_.scheduledAt)}")
     runService.startJobRun(spec)
-    scheduleNextRun()
+    scheduleNextRun(s)
   }
 
-  def scheduleNextRun(): Unit = {
+  def scheduleNextRun(s: ScheduleSpec): Unit = {
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: lets be more descriptive than `s`

File: jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala [hunk 7]
```diff
@@ -24,63 +25,91 @@ class JobSpecSchedulerActor(
...
+          cancelSchedule(updateableSchedule)
+          scheduleNextRun(newSchedule)
+        }
+      }
+    }
   }
 
-  def runJob(): Unit = {
-    log.info(s"Start next run of job ${spec.id}, which was scheduled for $scheduledAt")
+  def runJob(s: ScheduleSpec): Unit = {
```
Comment (kensipe) [inline jobs/src/main/scala/dcos/metronome/jobspec/impl/JobSpecSchedulerActor.scala]: lets be more descriptive than `s`

**Comment by Teots** [comment]
@kensipe Thanks for your feedback

**Comment by kensipe** [comment]
may reopen... waiting on product to prioritize

**Pull request closed without merging**